src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Android.cs

```diff
@@ -0,0 +1,390 @@
+        * This is needed in order to read Android v4.3 tzdata files.
+        *
+        * Android 10+ moved the up-to-date tzdata location to a module updatable via the Google Play Store and the
+        * database location changed (https://source.android.com/devices/architecture/modular-system/runtime#time-zone-data-interactions)
+        * The older locations still exist (at least the `/system/usr/share/zoneinfo` one) but they won't be updated.
+        */
+        private sealed class AndroidTzData
+        {
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataHeader
+            {
+                public fixed byte signature[12]; // "tzdata2012f\0"
+                public int indexOffset;
+                public int dataOffset;
+                public int finalOffset;
+            }
+
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataEntry
+            {
+                public fixed byte id[40];
+                public int byteOffset;
+                public int length;
+                public int unused; // Was raw GMT offset; always 0 since tzdata2014f (L).
+            }
+
+            private string[]? _ids;
+            private int[]? _byteOffsets;
+            private int[]? _lengths;
+
+            public AndroidTzData()
+            {
+                ReadHeader(GetTimeZoneDirectory() + _timeZoneFileName);
+            }
+
+            private unsafe void ReadHeader(string tzFilePath)
+            {
+                int size = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+                var buffer = new byte[size];
+                AndroidTzDataHeader header = ReadAt<AndroidTzDataHeader>(tzFilePath, 0, buffer);
+
+                header.indexOffset = NetworkToHostOrder(header.indexOffset);
+                header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+                // tzdata files are expected to start with the form of "tzdata2012f\0" depending on the year of the tzdata used which is 2012 in this example
+                sbyte* s = (sbyte*)header.signature;
+                string magic = new string(s, 0, 6, Encoding.ASCII);
+
+                if (magic != "tzdata" || header.signature[11] != 0)
+                {
+                    var b = new StringBuilder();
+                    for (int i = 0; i < 12; ++i) {
+                        b.Append(' ').Append(((byte)s[i]).ToString("x2"));
+                    }
+
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_BadTZHeader, tzFilePath, b.ToString()));
+                }
+
+                ReadIndex(tzFilePath, header.indexOffset, header.dataOffset, buffer);
+            }
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2087:UnrecognizedReflectionPattern",
+                Justification = "Implementation detail of Android TimeZone")]
```

Comment:
This is not a valid justification of suppressing this warning.

What line is this warning about? Something else is wrong here that we need to figure out.

Reply:
I am hitting this error when removing this line.
`
/Users/mdhwang/runtime_droid/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Android.cs(319,17): Trim analysis error IL2087: System.AndroidTzData.ReadAt<T>(Int64,Byte[]): 'structureType' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors', 'DynamicallyAccessedMemberTypes.NonPublicConstructors' in call to 'System.Runtime.InteropServices.Marshal.PtrToStructure(IntPtr,Type)'. The generic parameter 'T' of 'System.AndroidTzData.ReadAt<T>(Int64,Byte[])' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/mdhwang/runtime_droid/src/libraries/src.proj]
`

Reply:
Ah, the right thing to do there is to annotate `T` with `DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors`. But like I said elsewhere, I wouldn't use this approach.

FYI - `UnconditionalSuppressMessage` is a "last resort" approach to solving ILLink warnings. They shouldn't be used lightly, as it very easily could lead to trimming bugs.